### DISCUSSION TITLE
Dynamic composable import

### DISCUSSION THREAD

**Discussion by milelazar** [nuxt/nuxt#30312] [Questions] [answered]
Hi,

is there a way to exclude come composables from build bundle based on app config? I have a composable with event tracking functions and it should be enabled or disabled based on app config.

But even if disabled and not used, it is bundled. Here are some minimal examples I have tried

Bundled
```
const config = useAppConfig();
if (config.tracking) {
  const {tracking} = useTracking();
}
```

Also tried dynamic import but still ends up bundled
```
if (config.tracking) {
  const {tracking} = await import('@/composables/useTracking');
}
```

This works (not bundled) but beats the whole idea of having config option
```
const enabled = false;
if (enabled) {
  const {tracking} = useTracking();
}
```

**Comment by dsvgl** [comment]
You probably want to use the destructured `tracking` variable later on in your code. Doing the const assignment inside the if statement and then using `tracking` later on in setup or template will not work.

Try:
```ts
let myTracking;

const config = useAppConfig();

if (config.tracking) {
  const { tracking } = useTracking();
  myTracking = tracking;
}
```

**Reply by milelazar** [reply]
Already tried this. But this is not the problem. I want to tree shake the composable if disabled (set to false in app config). And it does not work since it is automatically imported and bundled no matter the config.

Bigger picture: I use layers and there are bunch of "base" components and composables. On some projects base component is used but composable is not needed. So I want to avoid bundle bloat with those unused/disabled composables

**Reply by dsvgl** [reply, marked as answer]
Understood.

We heavily use vite defines for dead code elimination. Maybe you can try with that.

```ts
// base layer

export default defineNuxtConfig({
  vite: {
    define: {
      __TRACKING_ENABLED__: JSON.stringify(true),
    },
  },
});

// set to false the same way in desired project. Nuxt will merge the layers.
/// ...
__TRACKING_ENABLED__: JSON.stringify(false),
/// ...
```

Then in your component:
```vue
<script setup lang="ts">
if (__TRACKING_ENABLED__) { // will resolve to `true` or `false` at build time -> dead code elimination
  const {tracking} = useTracking();
}
</script>
```

**Reply by milelazar** [reply]
This works like a charm. Thank you!
